### ISSUE TITLE
Improve `simoc.py`

### ISSUE THREAD

**Issue by ezio-melotti** [overthesun/simoc#204]
I've been working on further automating the setup of SIMOC through `simoc.py`.  This issue summarizes some design decision and issues I met along the way.

First of all, we have two use cases:
1. backend only
2. backend + frontend

We also want to support both Linux and MacOS.  MacOS is lower priority and should be easy to do once Linux is done.  Grant already did some work in that direction:
* #201 

So far the `simoc.py` script supported the scenario 1 (backend only), so one of the goals is supporting the scenario 2 (backend + frontend).  For that, the steps are roughly:
1. clone the two repos
2. install dependencies (requires log off/log in)
3. setup frontend
4. setup backend

In order to do that there are a number of changes needed:
- [x] Users need to either setup ssh or generate a PAT (personal access token) before being able to clone the repos.
  * this cannot be automated, but it can be documented (PR incoming)
  * eventually we should make simoc pip-installable
- [x] `docker` and `docker-compose` need to be installed and set up.
  * on Linux this cannot be fully automated and a log out/log in (or restart) is required (more on this later)
  * #205
- [x] the frontend should also check/install/set up `docker` since it is set up before the backend
  * https://github.com/overthesun/simoc-web/pull/174
- [x] the frontend needs a `setup` command that combines and automates a few of the existing steps and a command to run `run.sh` to install `npm` deps and download files before building
  * https://github.com/overthesun/simoc-web/pull/175

The last few items are covered in
- [x] overthesun/simoc-web#172

---

## `docker` setup issues

After installing `docker` (and `docker-compose`) through `apt`, it seems that:
* a `docker` group is created automatically
* the current user is NOT added to the group
* the current user is NOT logged in the group

These steps are required in order to run the `docker` commands (and `simoc.py`) without `sudo`.  Adding the user to the group is trivial: `sudo usermod -aG docker $USER` does it, as described in https://docs.docker.com/engine/install/linux-postinstall/

However, this doesn't take immediate effect, since the user needs to "log in" the group.  Automating this has proved difficult.  These are the solutions I explored:
* Running `newgrp docker` (as suggested in the link above) seems to work when executed directly from the shell, however this appears to open a new shell where the user is logged in the group.  Executing further commands in this sub-shell works, but once the sub-shell is closed, the parent process is still not logged in the group.  This doesn't work when called from `simoc.py` because on top of suspending the installation until you quit the sub-shell, the command only affects the sub-shell and all other commands executed outside of it will fail.
* The `os` module has some APIs (such as [`os.setgid`](https://docs.python.org/3/library/os.html#os.setgid)), but without `sudo` the function raises a `PermissionDeniedError`.  In addition, the change will only affect the current process, and the group id would have to be set on any further invocation of the `simoc.py` script, thus requiring `sudo`.
* Similarly, [`subprocess.run`](https://docs.python.org/3/library/subprocess.html#subprocess.Popen) accepts a `group` and `extra_groups` argument, but without `sudo` it also raises a `PermissionDeniedError`.
* Creating a separate script to install and set up `docker` and running it with `sudo` will allow the above functions to work, but in addition of adding an extra step to the setup process, it will also not work if those functions only affect the current process and not the global environment.  This might work if there is a function that permanently/globally logs the user in the group (but in that case we wouldn't need the separate script).
* After executing a command with `sudo` (e.g. the `apt install docker`), subsequent commands run with `sudo` work without asking for the password for a limited period of time.  This means that during the setup we could prefix the docker commands with `sudo` and if they are within that window it won't ask for the password again.  However all subsequent invocation of `simoc.py` will ask for the password, unless we detect if the user is logged in the group and conditionally add the `sudo` only if it's not (or we can try without and add `sudo` if the command fails).
* Logging out/logging in (or restarting the machine) seems to be the simpler and most reliable solution, but it requires user intervention and adds a step to the setup.

Unless we find a solution, I'm going to go with the last option, at least for the time being.

---

After these problems are fixed, a full install (backend + frontend) should include the following steps:
1. clone simoc and simoc-web
   * this needs ssh setup or PAT
2. run `python3 simoc-web.py setup`, which will
   * install `docker` and `docker-compose`
   * ask the user to log out/log in (or restart)
3. run `python3 simoc-web.py setup` again, which will
   * resume the installation
   * build the docker image
   * download simdata/3D files
   * build the frontend
   * copy it in the simoc repo
4. run `python3 simoc.py setup`, which will
   * do a bunch of things
   * start SIMOC on localhost:8080

If the frontend is not needed, in step 2 we can run `python3 simoc.py setup` and then skip step 3.
